internal/pkg/cli/deploy/workload.go

```diff
@@ -368,17 +387,73 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	// mutex for synchronizing access to the output map.
+	var mu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+
+	// create an array of printers, one for each container.
+	termPrinters := make([]*syncbuffer.TermPrinter, len(buildArgsPerContainer))
+
+	// create a context and an error group.
+	ctx := context.Background()
+	g, _ := errgroup.WithContext(ctx)
+
+	// counter for indexing the output buffers.
+	count := 0
+
+	// create a new cursor and hide it.
+	cursor := cursor.New()
+	cursor.Hide()
+	defer cursor.Show()
+
 	for name, buildArgs := range buildArgsPerContainer {
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
-		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
-		}
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
+		// create a pipe for streaming the build and push output.
+		pr, pw := io.Pipe()
+
+		termPrinter := syncbuffer.NewTermPrinter(log.DiagnosticWriter)
+		termPrinters[count] = termPrinter
+		count++
+
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push image: %w", err)
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			out.ImageDigests[name] = ContainerImageIdentifier{
+				Digest:            digest,
+				CustomTag:         d.image.CustomTag,
+				GitShortCommitTag: d.image.GitShortCommitTag,
+			}
```

Comment:
nit: not a huge difference, but I feel like it's probably safe to unlock the map before `return nil` !
```suggestion
			mu.Lock()
			out.ImageDigests[name] = ContainerImageIdentifier{
				Digest:            digest,
				CustomTag:         d.image.CustomTag,
				GitShortCommitTag: d.image.GitShortCommitTag,
			}
			mu.Unlock()
```